### ISSUE TITLE
[MLT] bootstrap samples differ

### ISSUE THREAD

**Issue by wahn** [wahn/rs_pbrt#48]
Since commit 87b0b5ecb94aaada973381bde8e9497eb9eeb14d we can generate **bootstrap paths** multi-threaded. While debugging something else I figured that the generated **bootstrap samples** differ slightly between the C++ and Rust version. To dig deeper into the problem I modified both versions to print out very specialized information regarding the test scene (`assets/scenes/caustic.pbrt`) and the resulting bootstrap samples. 

In Rust (`src/core/sampling.rs`):

```rust
impl Distribution1D {
    pub fn new(f: Vec<Float>) -> Self {
        let n: i32 = f.len() as i32;
        // compute integral of step function at $x_i$
        let mut cdf: Vec<Float> = Vec::with_capacity(n as usize + 1);
        cdf.push(0.0 as Float);
        for i in 1..(n + 1) {
...
            if n == 600000 && f[(i - 1) as usize] > 0.0 as Float {
                println!("f[{:?}] = {1:2.8}", i - 1, f[(i - 1) as usize]);
            }
        }
...
    }
...
}
```

In C++ (`src/core/sampling.h`):

```c++
struct Distribution1D {
    // Distribution1D Public Methods
    Distribution1D(const Float *f, int n) : func(f, f + n), cdf(n + 1) {
        // Compute integral of step function at $x_i$
        cdf[0] = 0;
        for (int i = 1; i < n + 1; ++i) {
          if (n == 600000 && f[i - 1] > 0.0) {
            printf("f[%d] = %2.8f\n", i - 1, f[i - 1]);
          }
...
    }
...
};
```

I will **attach** the resulting difference between both versions to the issue and report on any progress ...

**Comment by wahn** [comment]
Here is the diff file ([diff.txt](https://github.com/wahn/rs_pbrt/files/2047685/diff.txt)) mentioned above:

```diff

243c243
< f[2471] = 0.00001476
---
> f[2471] = 0.00001474
344c344
< f[3823] = 0.10070046
---
> f[3823] = 0.05951537
432c432
< f[4906] = 0.00000642
...
```

**Comment by wahn** [comment]
I think I might have found the problem.

In C++ there is a piece of code to correct subpath sampling densities for infinite area lights:

```c++
int GenerateLightSubpath(
...
    // Correct subpath sampling densities for infinite area lights
    if (path[0].IsInfiniteLight()) {
        // Set spatial density of _path[1]_ for infinite area light
        if (nVertices > 0) {
            path[1].pdfFwd = pdfPos;
            if (path[1].IsOnSurface())
                path[1].pdfFwd *= AbsDot(ray.d, path[1].ng());
        }

        // Set spatial density of _path[0]_ for infinite area light
        path[0].pdfFwd =
            InfiniteLightDensity(scene, lightDistr, lightToIndex, ray.d);
    }
    return nVertices + 1;
}
```

In Rust I split that code. In `random_walk(...)` I deal with the **second** vertex (which gets created there):

```rust
pub fn random_walk<'a>(
    scene: &'a Scene,
    ray: &mut Ray,
    sampler: &mut Sampler,
    beta: &mut Spectrum,
    pdf: Float,
    max_depth: u32,
    mode: TransportMode,
    path: &mut Vec<Vertex<'a, 'a, 'a>>,
    density_info: Option<Float>,
) -> usize {
...
    // correct subpath sampling densities for infinite area lights
    if let Some(pdf_pos) = density_info {
        // set spatial density of _path[1]_ for infinite area light
        if bounces > 0 {
            path[1].pdf_fwd = pdf_pos;
            if path[1].is_on_surface() {
                path[1].pdf_fwd *= vec3_abs_dot_nrm(&ray.d, &path[1].ng());
            }
        }
        // set spatial density of _path[0]_ for infinite area light
        // path[0].pdf_fwd = infinite_light_density(scene, light_distr, &ray.d);
        // WARNING: Line above is done in generate_light_subpath() !!!
    }
    assert!(
        bounces + 1 == path.len(),
        "bounces = {:?}, path.len = {:?}",
        bounces,
        path.len()
    );
    bounces
}
```

Because the **first** (light) vertex gets created in `generate_light_subpath(...)` I deal with it there:

```rust
pub fn generate_light_subpath<'a>(
    scene: &'a Scene,
    sampler: &mut Sampler,
    max_depth: u32,
    time: Float,
    light_distr: &Arc<Distribution1D>,
    // TODO: light_to_index
    path: &mut Vec<Vertex<'a, 'a, 'a>>,
) -> usize {
...
    if pdf_pos == 0.0 as Float || pdf_dir == 0.0 as Float || le.is_black() {
        return 0_usize;
    }
    if let Some(light_pdf) = light_pdf {
        // generate first vertex on light subpath and start random walk
        let vertex: Vertex = Vertex::create_light(light, &ray, &n_light, &le, pdf_pos * light_pdf);
        let is_infinite_light: bool = vertex.is_infinite_light();
        path.push(vertex);
        let mut beta: Spectrum =
            le * nrm_abs_dot_vec3(&n_light, &ray.d) / (light_pdf * pdf_pos * pdf_dir);
        // println!(
        //     "Starting light subpath. Ray: {:?}, Le {:?}, beta {:?}, pdf_pos {:?}, pdf_dir {:?}",
        //     ray, le, beta, pdf_pos, pdf_dir
        // );
        if is_infinite_light {
            // set spatial density of _path[1]_ for infinite area
            // light is done in random_walk !!!
            n_vertices = random_walk(
...
            );
            // set spatial density of _path[0]_ for infinite area light
            path[0].pdf_fwd = infinite_light_density(scene, light_distr, &ray.d);
        } else {
            n_vertices = random_walk(
...
            );
        }
    }
    n_vertices + 1
}
```

The problem seems to be that the **ray direction is wrong** at that point:

```gdb
(gdb) p ray.d
$75 = {x = 0.357894659, y = 0.542428493, z = -0.760054469}
```

Above is the ray direction in the C++ code, below in the Rust code:

```gdb
(gdb) p ray.d
$3 = Vector3<f32> = {x = 0.671148658, y = 0.338981152, z = 0.659280658}
```

A **possible solution** would be to pass the ray direction (from `random_walk(...)`) back to the caller.

**Comment by wahn** [comment]
Commit b9d28f337c80c3498ad61c6fccd341145b325038 avoids the split and moves the Rust code where it belongs. The signature of `random_walk(...)` simplifies that way, but the bug ist still there and has to be solved.

**Comment by wahn** [comment]
Commit c293f6f3814bd3f2826f1307cfd369ffcd03547a brings the difference between C++ and Rust down to **16** (out of 600.000) **values**. See 
[diff2.txt](https://github.com/wahn/rs_pbrt/files/2054789/diff2.txt):

```diff
27665c27665
< 27664: 27.03012657
---
> 27664: 23.12003517
36306c36306
< 36305: 0.07716168
---
> 36305: 0.07200304
59885c59885
< 59884: 29.40234375
---
> 59884: 24.83382034
109619c109619
< 109618: 16.07372665
---
> 109618: 14.07734680
130596c130596
< 130595: 19.96478653
---
> 130595: 18.95233536
162233c162233
< 162232: 1.87913811
---
> 162232: 1.68257380
173579c173579
< 173578: 7.23545122
---
> 173578: 6.17441940
175745c175745
< 175744: 0.32102531
---
> 175744: 0.30594379
240785c240785
< 240784: 12.37079430
---
> 240784: 11.31699562
257580c257580
< 257579: 20.30904007
---
> 257579: 19.51568604
338813c338813
< 338812: 0.06317624
---
> 338812: 0.06304591
361133c361133
< 361132: 2.20274544
---
> 361132: 2.03493810
487769c487769
< 487768: 0.30640954
---
> 487768: 0.30018875
531198c531198
< 531197: 5.15708637
---
> 531197: 5.01058817
541217c541217
< 541216: 13.80323696
---
> 541216: 12.21747684
554436c554436
< 554435: 11.42179012
---
> 554435: 8.51139927
```

**Before** we had 240 (out of 600.000) differing values.

**Comment by wahn** [comment]
Something seems to go wrong regarding the **MIS weight** calculations:

```
# C++
(gdb) b mlt.cpp:192 if rngIndex == 27664
(gdb) r
(gdb) b mlt.cpp:130
bdpt.cpp:529
(gdb) p misWeight
$11 = 0.997996211
# Rust
(gdb) b /home/jan/git/self_hosted/Rust/pbrt/src/lib.rs:612 if w == 27664
(gdb) r
(gdb) b mlt.rs:292
bdpt.rs:2134
(gdb) p mis_weight_flt
$13 = 0.853629291
```

**Comment by wahn** [comment]
Is it on surface, or not?

```
# C++
bdpt.h:333
if (next.IsOnSurface()) returns true
(gdb) display pdf
before: pdf = 0.147163153
(gdb) n
after : pdf = 0.124431834
# Rust
bdpt.rs:559
if next.is_on_surface() returns false
p pdf
before: pdf = 0.147163153
after : pdf = 0.147163153
```

**Comment by wahn** [comment]
Commit 0cf23ad1bd3aaea8e367f4ad4a207ef9d043d439 fixes the issue, all 600.000 values are exactly the same now. No difference between Rust and C++ ...
